Remove prelude assert from ctors & dtors

thewilsonator opened the pull request:
Removes prelude `assert(this !is null, "null this");` from constructors & destructors.
Trying to assign or access any members would crash anyway.

andralex commented:
To reduce the size of each function, shouldn't the assert be massaged within the invariant if it exists?

thewilsonator commented:
Good idea! Not sure how to do this though, too tired. Will look into it tomorrow morning.

thewilsonator commented:
(The diff looks weird but thats just deindeting.)

After looking at the code more closely I realised that in oder to trigger this you had to construct or destruct an aggregate at the null memory location, which would crash if you were to access any members, which is required for a constructor or destructor to do anything useful other than logging (which would likely pick up the null pointer anyway).

This removes that assert so there is no need to move it to the invariant.

thewilsonator commented:
@WalterBright @yebblies This works now.

WalterBright commented on src/ddmd/func.d:
```diff
@@ -1388,18 +1388,18 @@ extern (C++) class FuncDeclaration : Declaration
             // Precondition invariant
             Statement fpreinv = null;
-            if (addPreInvariant())
+            if (addPreInvariant() && isCtorDeclaration() is null)
```
`addPreInvariant()` is a virtual function. Add a definition for it in `CtorDeclaration` for it instead that returns `false`.

thewilsonator replied:
Actually its already there!

WalterBright commented on src/ddmd/func.d:
```diff
@@ -1388,18 +1388,18 @@ extern (C++) class FuncDeclaration : Declaration
             // Postcondition invariant
             Statement fpostinv = null;
-            if (addPostInvariant())
+            if (addPostInvariant() && isDtorDeclaration() is null)
```
ditto

thewilsonator committed:
Remove prelude `assert(this, "null this");` from ctors & dtors

To trigger this you had to construct or destruct an aggregate at the
null memory location, which would crash if you were to access any
members, which is required for a constructor or destructor to do
anything useful other than logging (which would likely pick up the null
pointer anyway).

andralex commented on changelog/removePreludeAssert.dd:
```diff
@@ -0,0 +1,3 @@
+Removed prelude assert for constructors & destructors
+
+The compiler used to insert an `assert(this,”null this”);` at the start of constructors & destructors. To trigger these asserts you had to construct or destruct an aggregate at the null memory location. This would crash if you were to access any members, which is required for a constructor or destructor to do anything useful other than logging (which would likely pick up the null pointer anyway).
```
s/this,/this, /
s/you had to construct/one needed to construct/
s/if you were to access any members/upon any data member access/
s/ other than logging (which would likely pick up the null pointer anyway)//

andralex commented on src/ddmd/func.d:
```diff
@@ -3864,71 +3864,48 @@ extern (C++) class FuncDeclaration : Declaration
+
+        //e = new DsymbolExp(Loc(), inv);
+        //e = new CallExp(Loc(), e);
+        //e = e.semantic(sc2);
```
Remove commented-out code, or precede with a comment describing its usefulness.

thewilsonator replied:
That was there before (it only shows up in the diff because it is de-indented). I can do that, but I'm not sure why it was there in the first place or what purpose it originally served (illustrative?).

andralex replied:
no matter then

andralex reviewed (commented):
lgtm

yebblies commented on changelog/removePreludeAssert.dd:
```diff
@@ -0,0 +1,5 @@
+Removed prelude assert for constructors & destructors
+
+The compiler used to insert an `assert(this, ”null this”);` at the start of constructors & destructors.
```
Those don't appear to be ascii double quotes.

rainers commented on changelog/removePreludeAssert.dd:
```diff
@@ -0,0 +1,5 @@
+Removed prelude assert for constructors & destructors
+
+The compiler used to insert an `assert(this, "null this");` at the start of constructors & destructors.
+To trigger these asserts one needed to construct or destruct an aggregate at the null memory location. 
```
DAutotest complains about  trailing white space in this line.

ibuclaw commented:
I'm fine with this. @WalterBright has already approved.

BBasile commented:
You should have added opCast too. [There's an interesting case showing why here](https://forum.dlang.org/post/ot20f2$2gub$1@digitalmars.com).
